crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -74,6 +75,147 @@ impl SemanticSyntaxChecker {
+                    Self::add_error(
+                        ctx,
+                        SemanticSyntaxErrorKind::WriteToDebug(WriteToDebugKind::Store),
+                        name.range,
+                    );
+                }
+                if let Some(type_params) = type_params {
+                    for type_param in type_params.iter() {
+                        if type_param.name() == "__debug__" {
+                            Self::add_error(
+                                ctx,
+                                SemanticSyntaxErrorKind::WriteToDebug(WriteToDebugKind::Store),
+                                type_param.name().range,
+                            );
+                        }
+                    }
+                }
+                for parameter in parameters {
+                    if parameter.name() == "__debug__" {
+                        Self::add_error(
+                            ctx,
+                            SemanticSyntaxErrorKind::WriteToDebug(WriteToDebugKind::Store),
+                            parameter.name().range,
+                        );
+                    }
+                }
+            }
+            Stmt::ClassDef(ast::StmtClassDef {
+                name, type_params, ..
+            }) => {
+                // test_err debug_shadow_class
+                // class __debug__: ...  # class name
+                // class C[__debug__]: ...  # type parameter name
+                if name == "__debug__" {
+                    Self::add_error(
+                        ctx,
+                        SemanticSyntaxErrorKind::WriteToDebug(WriteToDebugKind::Store),
+                        name.range,
+                    );
+                }
+                if let Some(type_params) = type_params {
+                    for type_param in type_params.iter() {
+                        if type_param.name() == "__debug__" {
+                            Self::add_error(
+                                ctx,
+                                SemanticSyntaxErrorKind::WriteToDebug(WriteToDebugKind::Store),
+                                type_param.name().range,
+                            );
+                        }
+                    }
+                }
+            }
+            Stmt::TypeAlias(ast::StmtTypeAlias {
+                type_params: Some(type_params),
+                ..
+            }) => {
+                // test_err debug_shadow_type_alias
+                // type __debug__ = list[int]  # visited as an Expr but still flagged
+                // type Debug[__debug__] = str
+                for type_param in type_params.iter() {
+                    if type_param.name() == "__debug__" {
+                        Self::add_error(
+                            ctx,
+                            SemanticSyntaxErrorKind::WriteToDebug(WriteToDebugKind::Store),
+                            type_param.name().range,
+                        );
+                    }
```

Comment:
should we add a `check_identifier` function that handles the `if name == "__debug__"` handling to avoid some repetition?